tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java

```
@@ -291,4 +320,179 @@ public String toString() {
+        if (metaProperties == null) {
+            throw new TerseException("Unable to read meta.properties from " + metadataDirectory);
+        }
+        if (!metaProperties.directoryId().isPresent()) {
+            throw new TerseException("No directory id found in " + metadataDirectory);
+        }
+        return metaProperties.directoryId().get();
+    }
+
+    static Set<RaftVoterEndpoint> getControllerAdvertisedListeners(
+        Properties props
+    ) throws Exception {
+        Map<String, Endpoint> listeners = new HashMap<>();
+        SocketServerConfigs.listenerListToEndPoints(
+            props.getOrDefault(SocketServerConfigs.LISTENERS_CONFIG, "").toString(),
+            __ -> SecurityProtocol.PLAINTEXT).forEach(e -> listeners.put(e.listenerName().get(), e));
+        SocketServerConfigs.listenerListToEndPoints(
+            props.getOrDefault(SocketServerConfigs.ADVERTISED_LISTENERS_CONFIG, "").toString(),
+            __ -> SecurityProtocol.PLAINTEXT).forEach(e -> listeners.put(e.listenerName().get(), e));
+        if (!props.containsKey(KRaftConfigs.CONTROLLER_LISTENER_NAMES_CONFIG)) {
+            throw new TerseException(KRaftConfigs.CONTROLLER_LISTENER_NAMES_CONFIG +
+                " was not found. Is this a valid controller configuration file?");
+        }
+        LinkedHashSet<RaftVoterEndpoint> results = new LinkedHashSet<>();
+        for (String listenerName : props.getProperty(
+                KRaftConfigs.CONTROLLER_LISTENER_NAMES_CONFIG).split(",")) {
+            listenerName = ListenerName.normalised(listenerName).value();
+            Endpoint endpoint = listeners.get(listenerName);
+            if (endpoint == null) {
+                throw new TerseException("Cannot find information about controller listener name: " +
+                    listenerName);
+            }
+            results.add(new RaftVoterEndpoint(endpoint.listenerName().get(),
+                    endpoint.host() == null ? "localhost" : endpoint.host(),
```

Should it be an error if the host is `null`? Do we assume localhost for brokers if they don't specify a host in the listeners or advertise listeners configuration?

---

I was on the fence about this.

I think for local testing, I can definitely imagine some people putting `CONTROLLER://:9093` in the advertised listeners, and not thinking anything more about it. This would certainly be a bad idea in production, though.

We could throw an exception in that situation, or we could translate it to `localhost`... Perhaps the exception is better?